rfcs/2022-02-23-9481-config-schema.md

```diff
@@ -0,0 +1,710 @@
+
+Incremental steps to execute this change. These will be converted to issues after the RFC is
+approved:
+
+- [ ] Develop the `Configurable` trait, and supporting types, and implement it by hand for Rust
+  standard library types, and at least one of each major component type (source, transform, sink) to
+  vet out the design and expose any corner cases.
+- [ ] Develop an initial set of helper types/methods for generating a JSONSchema document from a
+  root `Configurable` type, and implement the `JSONSchema` trait from `schemars` by hand for all
+  types which have a manual `Configurable` implementation.
+- [ ] Create a top-level root type that fully encompasses the concept of a Vector configuration, and
+  manually implement the `JSONSchema` trait for it.
+- [ ] Create a Vector subcommand for running the schema generation code that prints it to the
+  console.
+- [ ] Develop an initial version of the `Configurable` derive macro that can generate a boilerplate
+  `Configurable` trait implementation without any support for attributes or attribute interrogation.
+- [ ] Add macro support for also deriving the boilerplate `JSONSchema` trait implementation.
+- [ ] Add macro support for interrogating `serde` type/field/variant attributes to generate metadata
+  with a goal of being on par with what `schemars` supports, plus whatever they don't support that
+  we need.
+- [ ] Add macro support for `configurable` attributes, specifically in the vein of what `schemars`
+  provides, in terms of adding field value constraints or custom metadata.
+- [ ] Replace hand-written implementations of `Configurable`/`JSONSchema` with the derive
+  macro/attributes.
+- [ ] Continue updating configuration types with the derive macro until all components are using it.
+- [ ] Update the logic used to register a component (the `inventory`-based stuff) to enforce that
+  the configuration types implement `Configurable`, thereby ensuring that all configuration types
+  are adhering to the requirements enforced by the derive macro.
+
+## Future Improvements
+
+- Move all configuration validation to `configurable` attributes, as well as default values that get
+  merged in after-the-fact. Currently, many defaults are merged in, and validations are performed,
+  only once a sink is in the process of being built, and the configuration has been deserialized,
+  which means some validation happens during deserialization and some happens after, leading to
+  slightly discontiguous error messages.
+- Generate a configuration schema whenever we cut a release, and store it in Git, followed by
+  checking it against the last release's schema to check for incompatibilities: removal of fields
+  that weren't already marked as deprecated, etc.
+- Generate Cue definitions/documentation based on the JSONSchema itself. This is apparently a
+  one-liner `cue` command but would also involve figuring out how to integrate the resulting Cue
+  into our existing Cue documentation.
```

Comment:
Would it be worth while just keeping it as json and moving off of cue?

Reply:
Possibly, but I think that would fall under a much bigger umbrella of how the website is developed, and the pathway from JSONSchema to Cue seems far easier in terms of being able to hand it off for integration.

I'm sure @binarylogic has opinions here, but my guess is that there will be another RFC-esque process for any change that big, and they'll figure out what approach they want to take when that happens, and not a second before, given the amount of work to do so.

Reply:
Yeah, mostly thinking that the cue is rendered to JSON to actually be used by the site - so that transition seems natural (though out of scope to this RFC)

Reply:
Yes, we will follow up this RFC with a variety of enhancements that build on this work. The ultimate goal is to get rid of cue entirely. @jdrouet is a big fan of cue, so I'll need to convince him. Otherwise, that's the plan.